crates/core/src/transport/udp_transport.rs

```diff
@@ -0,0 +1,181 @@
+//! UDP transport implementation
+//!
+//! # Protocol
+//!
+//! ## Connection Establishment
+//!
+//! ### Neither peer is a gateway
+//!
+//!
+//!
+//! # Transport message types (u8)
+//!
+//! * 0: Symmetric key encrypted with our public key
+//! * 1: Acknowledgement of symmetric key - encrypted with symmetric key
+//! * 2: Message - encrypted with symmetric key
+//! * 3: Disconnect message - encrypted with symmetric key
+
+use super::*;
+use crate::transport::crypto::{TransportKeypair, TransportPublicKey};
+use crate::transport::udp_connection::UdpConnection;
+use aes::cipher::KeyInit;
+use aes::Aes128;
+use dashmap::DashMap;
+use rand::random;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::vec::Vec;
+use tokio::net::UdpSocket;
+use tokio::sync::{mpsc, RwLock};
+use tokio::task;
+use tracing::Value;
+
+/// The maximum size of a received UDP packet, MTU is 1500
+/// so this should be more than enough.
+const MAX_PACKET_SIZE: usize = 2048;
+
+const RECEIVE_QUEUE_SIZE: usize = 100;
+const SEND_QUEUE_SIZE: usize = 100;
+
+pub(crate) struct UdpTransport {
+    pub(super) connections: DashMap<SocketAddr, UdpConnection>,
+    keypair: TransportKeypair,
+    listen_port: u16,
+    is_gateway: bool,
+    max_upstream_rate: BytesPerSecond,
+    send_queue: mpsc::Sender<(SocketAddr, Vec<u8>)>,
+}
+
+impl UdpTransport {
+    async fn new(
+        keypair: TransportKeypair,
+        listen_port: u16,
+        is_gateway: bool,
+        max_upstream_rate: BytesPerSecond,
+    ) -> Result<Arc<RwLock<Self>>, TransportError>
+    where
+        Self: Sized,
+    {
+        // Bind the UDP socket to the specified port
+        let socket = UdpSocket::bind(("0.0.0.0", listen_port))
+            .await
+            .map_err(|e| TransportError(e.to_string()))?;
+
+        let (send_queue, mut send_queue_receiver) = mpsc::channel(SEND_QUEUE_SIZE);
+
+        let new_transport = Arc::new(RwLock::new(UdpTransport {
+            connections: DashMap::new(),
+            keypair,
+            listen_port,
+            is_gateway,
+            max_upstream_rate,
+            send_queue,
+        }));
+
+        // Spawn a task for listening to incoming UDP packets
+        let new_transport_clone = Arc::clone(&new_transport);
+        task::spawn(async move {
```

Comment:
I don't think we should hide this here... let the consumers of the code handle those connections however they want and we simply stick to establishing a reliable connection in this code.